### ISSUE TITLE
Re-entering zones through recursion inflates total time in zone statistics

### ISSUE THREAD

**Issue by rokopt** [wolfpld/tracy#159]
If a `ZoneScoped` surrounds a recursive call, the time spent in child zones of the recursive call appears to be accounted to the parent once for each occurrence of the parent on the stack.  For example, if I run the following (Catch-2) test:

```
static void
child()
{
    ZoneScoped;
    sleep(1);
}

static void
parent(size_t numRecursions)
{
    ZoneScoped;
    if (numRecursions == 0)
    {
        child();
    }
    else
    {
        parent(numRecursions - 1);
    }
}

TEST_CASE("Tracy recursive zones")
{
    sleep(5); // allow time for Tracy handshake
    parent(10);
}
```

When I bring up "Statistics", the "self time" shows what I'd expect -- 1s in `child()` and practically nothing in `parent()` -- but when I un-check "self time" to show the total time, `parent()` shows 11s.

**Comment by wolfpld** [comment]
This is how inclusive time counting works, which you explicitly requested by disabling the "self time" option.

**Comment by rokopt** [comment]
Thanks for the reply.  I'll close the issue.

**Comment by rokopt** [comment]
Is there some other statistic I could examine, or zone annotation that I could use, which would show ~1s for `parent()` in this case?  I'd like to know what latencies clients are getting when they call a function such as `parent()`, whose implementation might make some (possibly indirect) recursive calls which would induce the inclusive zone time time to reflect more than the wall clock time spent in the call from the client (and whose "self time" does not reflect the latency either because much of the time is spent in child zones).

**Comment by wolfpld** [comment]
You can inspect the "child zones" list or "time distribution" for the parent zone in the zone information window.

**Comment by rokopt** [comment]
> You can inspect the "child zones" list or "time distribution" for the parent zone in the zone information window.

Thanks.  The "child zones" list does show 1s in `parent`.  I think that view is only for single invocations, though -- there's not a summary form of that one, is there?

And (sorry to keep at this) the total time displayed in the "time distribution" for `parent` is also 11s -- is it not a bug for a single-zone time accumulator to show a time greater than the wall-clock time spent by a single-threaded program?  (It occurs to me that if I'd called `sleep()` directly from `parent()` in my example above instead of calling `child()`, then the `parent` zone would have a self time of 1s and an inclusive time of 11s.  If there's only one zone in the entire system, shouldn't the self and inclusive time for that one zone be equal?)

I'm imagining writing a codebase using, for example, some recursive red-black tree algorithms like those in many `std::map` implementations.  If I saw a high inclusive time for some of them, and replaced them with versions that I thought would be faster and which happened to be iterative, I might see the inclusive times for their zones drop way down the "Statistics" list, and think I'd succeeded -- even if my new algorithms weren't really faster, and their inclusive times had only gone down because they weren't being multiplied by recursion depth anymore.

**Comment by rokopt** [comment]
> I'm imagining writing a codebase using, for example, some recursive red-black tree algorithms like those in many std::map implementations. If I saw a high inclusive time for some of them, and replaced them with versions that I thought would be faster and which happened to be iterative, I might see the inclusive times for their zones drop way down the "Statistics" list, and think I'd succeeded -- even if my new algorithms weren't really faster, and their inclusive times had only gone down because they weren't being multiplied by recursion depth anymore.

I might look at the code and try to figure out how to add an accumulator that doesn't count re-entered zones and thus doesn't have the above property.  If I managed to, might you be interested in a PR?

**Comment by wolfpld** [comment]
I'd like you to first describe what the intended result would be. I think you have some intuition on how this should be working, but it was not explicitly specified here.

**Comment by rokopt** [comment]
I'd like a "self plus child time" zone statistic that does not increment its call count or accumulate time when a thread re-enters a zone.  Unlike the existing non-"self time" statistic, it would count only the first appearance of any one zone on a thread's call stack.

In the example code above, the `parent` zone's new statistic would show `1` call for `1s`.  (Currently, `parent`'s "self time" shows `11` calls for negligible time, and its non-"self time" shows `11` calls for `~11s`.)

I'd be inclined to make that the default behavior of non-"self time" because I find it less surprising than the current behavior:  I'd expect to see similar numbers for iterative and recursive implementations of the same interface if their practical performance were similar; I wouldn't expect to see much higher numbers for the recursive implementation just because it happened to be recursive.  But if you prefer, the accumulator that doesn't count recursive calls could be a new one ("non-reentrant time"?  There is likely a better name) and the "self time" checkbox could become a tri-state.

Thanks again for your replies!

**Comment by wolfpld** [comment]
Ok, I think "count only the first appearance of any one zone on a thread's call stack" is a good description for this feature. The tri-state approach, with the proposed solution being available as a new option would be the way to go.

My only concern here would be performance implications - do note that statistics are updated live during the capture and that you can have one top-level zone, with billion (indirect) children. This is not an exaggeration:

![obraz](https://user-images.githubusercontent.com/600573/104828899-59225080-586e-11eb-8822-b16e965a907e.png)

**Comment by rokopt** [comment]
> Ok, I think "count only the first appearance of any one zone on a thread's call stack" is a good description for this feature. The tri-state approach, with the proposed solution being available as a new option would be the way to go.

Great, thanks again.  I'll work on that.

**Comment by rokopt** [comment]
Reopening to work on this.

**Comment by rokopt** [comment]
I drafted some initial changes which are nowhere near complete, but which I wonder whether you'd be willing to comment on to confirm or deny that the general approach is sound, before I wrote any more code in a potentially nonsensical direction.  :)  (This is the first time I've looked at Tracy code, so that's quite likely!)  The changes are at https://github.com/stellar/tracy/pull/2/commits/6c96b1e2b5f319c0bf9da37a9b90dac44579edf2 and some discussion is at https://github.com/stellar/tracy/pull/2#issuecomment-767221310 .

Facets in which it currently is or might be incomplete, and some questions:

- Regarding your performance concern, the changes currently detect re-entries of zones by using `std::find_if()` on `ThreadData::stack`.  I suspect that this linear search of a stack that could easily be dozens of frames deep in a method called as often as `Worker::NewZone()` will be intolerable in some applications.  To begin with, I'd just like to know whether it's semantically correct.  If it is, then the next question would be how to make it fast enough.  I'm not sure whether, for example, adding something like an `unordered_map` mapping source locations to `ZoneEvent`s to `struct ThreadData` would suffice, and whether it would be okay to increase the size of `ThreadData` by doing that.
- Also regarding performance, it occurs to me that since we only care whether a zone is being re-entered when we're measuring non-reentrant time, we might be able to skip whatever test for re-entry that we end up with altogether when we're not measuring non-reentrant time.  I'm not sure whether that would require that the choice of whether to display non-reentrant time be something like a command-line option, or a global toggle, or whether it could be a choice in the "Statistics" panel and the time could be computed on the fly when that box was checked.
- And speaking of checking boxes, I haven't changed the GUI at all yet (for example, to replace the "Self time" toggle with a tri-state) -- right now the new behavior just takes over under an `#ifdef`, replacing the existing "total time" statistic.  This is obviously not a final form, but only to test the new behavior.  As we've discussed, a tri-state in the "Statistics" panel would be an option; others would include a command-line option or a global toggle.  I don't have much context for evaluating the trade-offs among these options.
- As noted in the [comment linked above](https://github.com/stellar/tracy/pull/2#issuecomment-767221310), I haven't touched the _count_ of zone entries or the display of zone entries in the "Find zone" panel (either the histogram or the zone list).  I think that even when we're displaying the new "non-reentrant" time, we probably still want to see all the zone entries, including the re-entries, in the "Find zone" details.  But I could be wrong.
- I'm currently tracking in the `ZoneEvent` structure whether the event is a re-entry.  Right now that's just a `bool` -- that can't be right in the long run for a tightly-packed structure, but I haven't yet even tried to determine whether there are any available bits already (or bits that could be made available).  I _think_ from my reading of the code that `ZoneEvent`s processed by `ProcessZoneEnd()`, `ReconstructZoneStatistics()`, and `DrawStatistics()` will all have gone through `NewZone()` and thus will have `reentry` set if necessary, but I could be missing something.  I also haven't spotted any other paths that would need to _check_ for `reentry`, but there too I might be missing some.

Thanks again for your time.  I hope the newbie questions aren't overly burdensome.  :)

**Comment by wolfpld** [comment]
> Regarding your performance concern, the changes currently detect re-entries of zones by using `std::find_if()` on `ThreadData::stack`.  I suspect that this linear search of a stack that could easily be dozens of frames deep in a method called as often as `Worker::NewZone()` will be intolerable in some applications.  To begin with, I'd just like to know whether it's semantically correct.  If it is, then the next question would be how to make it fast enough.  I'm not sure whether, for example, adding something like an `unordered_map` mapping source locations to `ZoneEvent`s to `struct ThreadData` would suffice, and whether it would be okay to increase the size of `ThreadData` by doing that.

`ThreadData` is a per-thread struct, so adding things to it is ok. I think using a map (`unordered_flat_map`) with `SrcLoc -> entry count` mapping would be sufficient here to make this work.

> Also regarding performance, it occurs to me that since we only care whether a zone is being re-entered when we're measuring non-reentrant time, we might be able to skip whatever test for re-entry that we end up with altogether when we're not measuring non-reentrant time. I'm not sure whether that would require that the choice of whether to display non-reentrant time be something like a command-line option, or a global toggle, or whether it could be a choice in the "Statistics" panel and the time could be computed on the fly when that box was checked.

There are multiple ways of how this calculation is being performed, for different usage scenarios. The `SourceLocationZones` struct stores global statistics, but if you want to limit the time range you have to iterate over the zones (`m_statRange.active` with  `m_statMode == 0`). There are also various caches to calculate this only when a change occurs (`m_statCache`).

> And speaking of checking boxes, I haven't changed the GUI at all yet (for example, to replace the "Self time" toggle with a tri-state) -- right now the new behavior just takes over under an `#ifdef`, replacing the existing "total time" statistic.  This is obviously not a final form, but only to test the new behavior.  As we've discussed, a tri-state in the "Statistics" panel would be an option; others would include a command-line option or a global toggle.  I don't have much context for evaluating the trade-offs among these options.

A tri-state (maybe as three radio buttons?) selection should be used.

> As noted in the [comment linked above](https://github.com/stellar/tracy/pull/2#issuecomment-767221310), I haven't touched the _count_ of zone entries or the display of zone entries in the "Find zone" panel (either the histogram or the zone list).  I think that even when we're displaying the new "non-reentrant" time, we probably still want to see all the zone entries, including the re-entries, in the "Find zone" details.  But I could be wrong.

The non-reentrant count in the statistics view should show only non-reentrant zones. Count of reentrant zones will be still available in other display modes.

As for the other views, *if* the find zone would show non-reentrant times, then only the non-reentrant zones should be included. Not sure if this is something I want to maintain, however, as each additional option doubles the number of cases to handle here and there's already too many.

> I'm currently tracking in the `ZoneEvent` structure whether the event is a re-entry.  Right now that's just a `bool` -- that can't be right in the long run for a tightly-packed structure, but I haven't yet even tried to determine whether there are any available bits already (or bits that could be made available).  I _think_ from my reading of the code that `ZoneEvent`s processed by `ProcessZoneEnd()`, `ReconstructZoneStatistics()`, and `DrawStatistics()` will all have gone through `NewZone()` and thus will have `reentry` set if necessary, but I could be missing something.  I also haven't spotted any other paths that would need to _check_ for `reentry`, but there too I might be missing some.

I'd advise to first look for a solution that doesn't add any data to the `ZoneEvent` struct.

**Comment by rokopt** [comment]
Thanks for the detailed comments.

> I'd advise to first look for a solution that doesn't add any data to the ZoneEvent struct.

In `ProcessZoneEnd()`, I think that would be straightforward -- we have the thread stack there, so we could check it the way that my current code does in `NewZone()` (in future iterations, using an `unordered_flat_map`).

In `ReconstructZoneStatistics()`, I think that we're processing each individual thread's entire sequence of events serially, so with no interleaving between events for different threads, perhaps we could maintain a `recursionDepth` in the `SourceLocationZones`?

Are those two ideas sensible?  (I haven't yet thought of an approach for the two range-active cases of `DrawStatistics()` that doesn't affect the `ZoneEvent` structure.)

**Comment by wolfpld** [comment]
Both `ProcessZoneEnd()` and `ReconstructZoneStatistics()` can use similar algorithm, as the processing order is the same, as you have observerd. And the range-limiting case is the hard one.

**Comment by rokopt** [comment]
> Both ProcessZoneEnd() and ReconstructZoneStatistics() can use similar algorithm, as the processing order is the same, as you have observerd.

Ah, thanks.  I'd only realized that `ReconstructZoneStatistics()` processed each thread's events separately; I'd thought that `ProcessZoneEnd()` might be able to interleave events from different threads.

> And the range-limiting case is the hard one.

I've had one thought about this.  I noticed there's a `GetZoneCallstack()` which we could call on a `ZoneEvent` in `DrawStatistics()`, and a `GetCallstack()` that we can call on the result of that, which returns an array of `CallstackFrameId`s.  If we iterated through the frames to check whether there were any besides the last frame (`end() - 1`) that represented the same zone as the last frame -- maybe if they have the same `data`? -- would that tell us whether that `ZoneEvent` represented a re-entry to a zone that was already on the stack?

**Comment by wolfpld** [comment]
Callstack data is optional and represents a different data set. What you probably want to do is walk the zone stack up by getting zone parents and compare source location identifiers.

**Comment by rokopt** [comment]
> What you probably want to do is walk the zone stack up by getting zone parents and compare source location identifiers.

Ah, thanks; is https://github.com/stellar/tracy/pull/2/commits/027b4c22c2dcf7cbf0fe444210529f25fa312965 a correct interpretation of the algorithm you mean?  (I haven't made any further changes in that commit, such as to get rid of the remaining uses of the field that I added to `ZoneEvent`; I figure I should check whether this is the right idea before I build on it.)

**Comment by wolfpld** [comment]
The basic implementation is sound, but it occured to me that it could be greatly improved. Each parent query involves walking through all the parent zones anyway, as all we have is a timestamp and the zone identifier itself, and we have to search if the zone is present in any of the threads, going down from the top level zones list in each. Your query can be thus improved, as with an appropriate modification of the parent-getter algorithm there won't be the need to go up one parent at a time.

For the time being, please work with what you have, as I think the parent-getter algorithm itself may not need to search across all threads. Give me some time to check on this, to save you the need to adapt the changes, if there would be any.

**Comment by wolfpld** [comment]
Ok, there is now a conditional fast path in `GetZoneParent()`.

**Comment by rokopt** [comment]
> Ok, there is now a conditional fast path in GetZoneParent().

Nice, thank you!  It looks to me as though the interface didn't change, so I don't think I have to change the loop that I wrote, correct?

In light of this optimization, do you think it would make sense to use that loop in `ProcessZoneEnd()` and `ReconstructZoneStatistics()` as well, or should I still make a new `unordered_flat_map` to use in those two places?

When it comes to plugging this in to the interface, I think at first glance that I should change `m_statSelf` from a `bool` to a tri-state `enum`, and add a new `nonReentrantTotal` field to `SrcLocZonesSlim`.  Is that sensible?

**Comment by wolfpld** [comment]
> I don't think I have to change the loop that I wrote, correct?

You still have to. Look closely at what `GetZoneParent()` and your loop do - this works, but makes little sense as it is.

> In light of this optimization, do you think it would make sense to use that loop in `ProcessZoneEnd()` and `ReconstructZoneStatistics()` as well, or should I still make a new `unordered_flat_map` to use in those two places?

Hash map will be faster.

> When it comes to plugging this in to the interface, I think at first glance that I should change `m_statSelf` from a `bool` to a tri-state `enum`, and add a new `nonReentrantTotal` field to `SrcLocZonesSlim`. Is that sensible?

Yes. At this point it may be better to use just one `total` value in `SrcLocZonesSlim` and update it as selected mode changes. This should be doable with proper cache invalidation (i.e. extend `if( cit != m_statCache.end() && cit->second.range == m_statRange && cit->second.sourceCount == it->second.zones.size() )`).

**Comment by rokopt** [comment]
> You still have to. Look closely at what GetZoneParent() and your loop do - this works, but makes little sense as it is.

Oh, maybe I see what you meant about each parent query having to walk all parents.  Perhaps I could create a versions of `GetZoneParent()` which perform the same algorithms as the two existing ones, but which instead of returning the eventual value of `parent`, it tests each `parent` to see if it has the same `SrcLoc()` as `zone` itself, and if so, returns `true`.  If it gets to the point where it would return a `parent` (or `nullptr`) without having returned `true`, then it returns `false`.  Is that the right idea?

If it is, should I search for a way of factoring out the common outline between the new and existing methods, or would it be clearer just to write new methods with similar outlines?

**Comment by wolfpld** [comment]
Yes, this is what I had in mind. Please don't trouble yourself with extracting the common code, as the implementations differ in little nuances and trying to make it a common handler would make it hard to read.

**Comment by rokopt** [comment]
Thank you!  Do the methods at https://github.com/rokopt/tracy/blob/49ca7b1b10d9c656735445dfd73c26dbbbbda98b/server/TracyWorker.cpp#L3201-L3286 look correct?

For the moment I put them in `Worker`, with forwarding methods from `View`.  The reason is so that I can call them from `Worker::ReconstructZoneStatistics()`.  I tried using `find_if()` in the latter in the same way as in `Worker::ProcessZoneEnd()` and it would crash when I loaded a trace, with a backtrace which looked as though `m_threadCtxData->stack` might not have been initialized yet.  I'm not sure whether that's something I'm misinterpreting, or something which would be fixed if I do the right thing in the unordered map implementation, which I haven't tried yet.  So I don't know whether `View` or `Worker` is the better place for the new `IsZoneReentry()` methods yet.

(I haven't made any of the other planned changes yet, such as adding an unordered map for `Worker::ProcessZoneEnd()` to search; I figure I should try to get one algorithm right at a time!)

**Comment by wolfpld** [comment]
> Thank you! Do the methods at https://github.com/rokopt/tracy/blob/49ca7b1b10d9c656735445dfd73c26dbbbbda98b/server/TracyWorker.cpp#L3201-L3286 look correct?

At a quick glance it looks fine.

>  I tried using `find_if()` in the latter in the same way as in `Worker::ProcessZoneEnd()` and it would crash when I loaded a trace, with a backtrace which looked as though `m_threadCtxData->stack` might not have been initialized yet.

The stack is not initialized (and won't ever be), as it is only used during live capture and not when loading a trace from disk.

> So I don't know whether `View` or `Worker` is the better place for the new `IsZoneReentry()` methods yet.

Definitely `View`, as what `IsZoneReentry()` does is appropriate for checking a single zone (and due to this you can't build acceleration structures, like the stack). What you should do instead is replicate the missing stack functionality when doing statistics reconstruction. This way you will have a handy way of checking the list of parent source locations, without the need to perform multiple binary searches for each zone.

**Comment by rokopt** [comment]
I'm sorry for the long delay!  I've finally opened a PR.

**Comment by rokopt** [comment]
Would you like me to close this as resolved by #230 , or should I leave that to you?

**Comment by wolfpld** [comment]
Either way works.
